operator/pkg/apis/operator/v1alpha1/defaults.go

```diff
@@ -0,0 +1,205 @@
+package v1alpha1
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+
+	"github.com/karmada-io/karmada/operator/pkg/constants"
+)
+
+var (
+	etcdImageImageRepository                   = fmt.Sprintf("%s/%s", constants.KubeDefaultRepository, constants.Etcd)
+	karmadaAPIServiceImageRepository           = fmt.Sprintf("%s/%s", constants.KubeDefaultRepository, constants.KarmadaAPIServer)
+	karmadaAggregratedAPIServerImageRepository = fmt.Sprintf("%s/%s", constants.KarmadaDefaultRepository, constants.KarmadaAggregatedAPIServer)
+	kubeControllerManagerImageRepository       = fmt.Sprintf("%s/%s", constants.KubeDefaultRepository, constants.KubeControllerManager)
+	karmadaControllerManagerImageRepository    = fmt.Sprintf("%s/%s", constants.KarmadaDefaultRepository, constants.KarmadaControllerManager)
+	karmadaSchedulerImageRepository            = fmt.Sprintf("%s/%s", constants.KarmadaDefaultRepository, constants.KarmadaScheduler)
+	karmadaWebhookImageRepository              = fmt.Sprintf("%s/%s", constants.KarmadaDefaultRepository, constants.KarmadaWebhook)
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// RegisterDefaults adds defaulters functions to the given scheme.
+// Public to allow building arbitrary schemes.
+// All generated defaulters are covering - they call all nested defaulters.
+func RegisterDefaults(scheme *runtime.Scheme) error {
+	scheme.AddTypeDefaultingFunc(&Karmada{}, func(obj interface{}) { SetObjectDefaultsKarmada(obj.(*Karmada)) })
+	return nil
+}
+
+// SetObjectDefaultsKarmada set defaults for karmada
+func SetObjectDefaultsKarmada(in *Karmada) {
+	setDefaultsKarmada(in)
+}
+
+func setDefaultsKarmada(obj *Karmada) {
+	setDefaultsHostCluster(obj)
+	setDefaultsKarmadaComponents(obj)
+}
+
+func setDefaultsKarmadaComponents(obj *Karmada) {
+	if obj.Spec.Components == nil {
+		obj.Spec.Components = &KarmadaComponents{}
+	}
+
+	setDefaultsEtcd(obj.Spec.Components)
+	setDefaultsKarmadaAPIServer(obj.Spec.Components)
+	setDefaultsKarmadaAggregratedAPIServer(obj.Spec.Components)
+	setDefaultsKubeControllerManager(obj.Spec.Components)
+	setDefaultsKarmadaControllerManager(obj.Spec.Components)
+	setDefaultsKarmadaScheduler(obj.Spec.Components)
+	setDefaultsKarmadaWebhook(obj.Spec.Components)
+}
+
+func setDefaultsHostCluster(obj *Karmada) {
+	if obj.Spec.HostCluster == nil {
+		obj.Spec.HostCluster = &HostCluster{}
+	}
+
+	hc := obj.Spec.HostCluster
+	if hc.Networking == nil {
+		hc.Networking = &Networking{}
+	}
+	if hc.Networking.DNSDomain == nil {
+		hc.Networking.DNSDomain = pointer.String(constants.KarmadaDefaultDNSDomain)
+	}
+}
+
+func setDefaultsEtcd(obj *KarmadaComponents) {
```

Comment:
Why not set default replicas for etcd?

Reply:
https://github.com/karmada-io/karmada/pull/3349

yes, only if the pr be merged, can the default replicas be set.

Reply:
After #3349 is merged, I think we can set replicas for etcd.

Reply:
Done. @Poor12